openshift-hack/update-kubensenter.sh

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "$KUBE_ROOT/hack/lib/init.sh"
+
+# Convert a path relative to $KUBE_ROOT to a real path
+localpath() {
+    realpath "$KUBE_ROOT/$1"
+}
+
+# Configuration for fetching this file, relative to this repository root
+ENVFILE=openshift-hack/kubensenter.env
+
+# The source of the file, relative to the remote repository root
+SOURCE=utils/kubensenter/kubensenter
+
+# The destination of the file, relative to this repository root
+DESTINATION=openshift-hack/images/hyperkube/kubensenter
+
+usage() {
+    source_env
+    echo "Usage:"
+    echo "  $0 [--to-latest]"
+    echo
+    echo "Updates the local copy of $DESTINATION as configured in $ENVFILE:"
+    echo "  REPO: $REPO"
+    echo "  COMMIT: $COMMIT"
+    echo
+    echo "Options:"
+    echo "  --to-latest (or env UPDATE_TO_LATEST=1)"
+    echo "    Update $ENVFILE to the latest commit or tag in $REPO configured by the TARGET entry"
+    echo "    (currently \"$TARGET\"), and synchronize to the updated commit."
+    echo "    - If TARGET resolves to a branch, pin to the latest commit hash from that branch"
+    echo "    - If TARGET resolves to a tag, pin to the latest tag that matches that pattern"
+    echo "    - TARGET may be a glob-like expression such as \"v1.1.*\" that would match any of the following:"
+    echo "        v1.1.0 v1.1.3 v1.1.22-rc1"
+    exit 1
+}
+
+source_env() {
+    source "$(localpath "$ENVFILE")"
+    # Intentionally global scope:
+    REPO=${REPO:-"github.com/containers/kubensmnt"}
+    COMMIT=${COMMIT:-"main"}
+    TARGET=${TARGET:-"main"}
+}
+
+update_env() {
+    local repouri latest refhash reftype refname newcommit patch
+    source_env
+    repouri=https://$REPO.git
+    echo "Updating to latest $TARGET from $repouri"
+
+    latest=$(git \
+                   -c "versionsort.suffix=-alpha" \
+                   -c "versionsort.suffix=-beta" \
+                   -c "versionsort.suffix=-rc" \
+                 ls-remote \
+                   --heads --tags \
+                   --sort='-version:refname' \
+                   "$repouri" "$TARGET" \
+             | head -n 1)
+    if [[ -z $latest ]]; then
+        echo "ERROR: No matching ref found for $TARGET"
+        return 1
+    fi
+    refhash=$(cut -f1 <<<"$latest")
+    reftype=$(cut -d/ -f2 <<<"$latest")
+    refname=$(cut -d/ -f3 <<<"$latest")
+
+    if [[ $reftype == "tags" ]]; then
+        echo "  Latest tag is $refname ($refhash)"
+        # Important: Use '...' around $refname so the resulting envfile is safe:
+        newcommit="COMMIT='$refname' # ($refhash)"
```

Comment:
A tag can have single-quotes:
```
% git tag "'foo'"   
% git push Miciah "'foo'":"'foo'"
[...]
To github.com:Miciah/origin.git             
 * [new tag]                 'foo' -> 'foo'
git push Miciah "'foo'":"'foo'"  1.87s user 0.20s system 17% cpu 11.892 total            
% git ls-remote Miciah '*foo*'
a946e2b9cd8acfdc73eac930b81623dc63f1a220        refs/tags/'foo'
```
So I still think that this is more correct:
```suggestion
        newcommit=$(printf 'COMMIT=%q # (%s)' "$refname" "$refhash")
```
Single-quotes in tag names is a really pathological case though.

Reply:
True!

Reply:
Done.  I did a bit of refactoring too to make it easier to test this function.  (test not included in this PR)